Add eslint as linter tool

JuanMaRuiz opened the pull request:
As discussed in the opened issue [jscs](https://github.com/jquery/sizzle/issues/438) is deprecated and jQuery is using eslint as linter tool. I have configured eslint as is in the jQuery project. But I have had to configure something in a special way: For example: I have marked as to skip the rule "no-nested-ternary"

```
  "rules": {
	"no-nested-ternary": "warn"
  }
```

because the style of coding of sizzle is not the same as the jQuery.

I have tried to extract to functions but I cannot name the functions with a clear name or because I think it could increase the complexity of the code. I have also tried to replace ternaries with `if`s but it increase the weight of the build.

Please, let me know I this PR is useful and/or needs to be updated.

Best regards.

fix: #438

mgol commented on Gruntfile.js:
```diff
@@ -133,43 +133,36 @@ module.exports = function( grunt ) {
+				// See https://github.com/sindresorhus/grunt-eslint/issues/119
+				quiet: true,
+				configFile: ".eslintrc",
+				fix: true
```
`configFile` & `fix` shouldn't be needed. Compare with the [jQuery equivalent](https://github.com/jquery/jquery/blob/6ced2639b567ab4a5089f38a8e90efedc12801bf/Gruntfile.js#L99-L114).

mgol commented on Gruntfile.js:
```diff
@@ -133,43 +133,36 @@ module.exports = function( grunt ) {
-					requireDotNotation: null
-				},
-				src: [ files.source ]
+				src: files.source // requireDotNotation: null
```
Let's remove those comments (here & in a few other places), there's no real value in providing obsolete JSCS option names.

JuanMaRuiz replied:
Oks

mgol commented on Gruntfile.js:
```diff
@@ -133,43 +133,36 @@ module.exports = function( grunt ) {
-					requireDotNotation: null
-				},
-				src: [ files.source ]
+				src: files.source // requireDotNotation: null
```
Is there any reason this needs to exist separately instead of merging `src` & `dev` into `all`?

mgol commented on Gruntfile.js:
```diff
@@ -133,43 +133,36 @@ module.exports = function( grunt ) {
+				src: files.source // requireDotNotation: null
+			},
+			dev: {
+				src: [ "src/**/*.js", "Gruntfile.js", "test/**/*.js", "build/**/*.js" ]
```
Why do you specify these paths instead of using previous versions, e.g.: `[ files.source, files.grunt, files.karma, files.speed, files.tests ]`?

JuanMaRuiz replied:
I have added the same configuration as [jQuery](https://github.com/jquery/jquery/blob/6ced2639b567ab4a5089f38a8e90efedc12801bf/Gruntfile.js#L99-L114) what do you think is the best approach?

JuanMaRuiz replied:
Changed.

mgol commented on speed/.eslintrc:
```diff
@@ -0,0 +1,11 @@
+{
+  "extends": "../.eslintrc-browser.json",
+  "rules": {
+  },
```
Remove `rules` since it's empty.

mgol commented on speed/.eslintrc:
```diff
@@ -0,0 +1,11 @@
+{
```
The files should be named `.eslintrc.json`, not `.eslintrc`. The latter is deprecated.

mgol commented on src/sizzle.js:
```diff
@@ -8,1150 +8,1198 @@
+( function( window ) {
+	"use strict";
+
+	var i,
```
Let's not change indentation. De-indent the whole file, please.

mgol replied:
@JuanMaRuiz This is not resolved, the whole file is still indented. This part is quite important, I can't review the rest of the PR before that is done since all of the lines are currently changed.

JuanMaRuiz replied:
I have changed indentation. I hope is ok now.

mgol commented on src/sizzle.js:
```diff
@@ -8,1150 +8,1198 @@
-			if ( list[i] === elem ) {
-				return i;
+( function( window ) {
+	"use strict";
```
@gibson042 Do you think it's OK we enable strict mode?

JuanMaRuiz replied:
eslint configuration force to use strict mode. We can disable (or configure as a warning) if you prefer.

mgol replied:
I guess strict mode is fine since we already use it like that in jQuery so it has to work. So I'm fine with keeping it (@gibson042 I'm working on an assumption that you'd agree).

mgol commented on src/sizzle.js:
```diff
@@ -8,1150 +8,1198 @@
+			var isValidHigh = high !== high || escapedWhitespace;
 
-		return results;
+			function getHighValue() {
```
Is adding this function necessary?

JuanMaRuiz replied:
Sorry I forgot to remove it. As I comment in the PR I have tried to replace ternary operators with functions or `if` and seems that I forgot to revert this change. Thanks, I will change it!

mgol commented on src/sizzle.js:
```diff
@@ -8,1150 +8,1198 @@
+
+// Optimize for push.apply( _, NodeList )
+	try {
+		push.apply( ( arr = slice.call( preferredDoc.childNodes ) ), preferredDoc.childNodes );
```
Is this change necessary?

mgol commented on src/sizzle.js:
```diff
@@ -8,1150 +8,1198 @@
+
+		// Support: Android<4.0
+		// Detect silently failing push.apply
+		arr[ preferredDoc.childNodes.length ].nodeType; // eslint-disable-line no-unused-expressions
```
A general rule: I'd prefer to use the `// eslint-disable-next-line rule-name` comments so that we don't increase line lengths.

JuanMaRuiz replied:
Ok! I'll do in that way

mgol commented on src/sizzle.js:
```diff
@@ -8,1150 +8,1198 @@
+						i = 0;
+
+					// Can't trust NodeList.length
+					while ( ( target[ j++ ] = els[ i++ ] ) ) {
```
Is this change necessary?

mgol commented on src/sizzle.js:
```diff
@@ -8,1150 +8,1198 @@
-					if ( nid === expando ) {
-						context.removeAttribute( "id" );
+						// Expand context for sibling selectors
+						newContext = rsibling.test( selector ) &&
```
Is this change necessary?

JuanMaRuiz replied:
Yep, the length is more than 100 chars

mgol replied:
I don't think it is... Please check again after you de-indent the file.

mgol commented on src/sizzle.js:
```diff
@@ -8,1150 +8,1198 @@
+
+		// ID filter and find
+		if ( support.getById ) {
+			Expr.filter.ID = function( id ) {
```
I think we were using quoted names here so that Google Closure Compiler doesn't mangle those names. If that's the case, we can't change them.

@gibson042 Can you confirm or deny?

mgol replied:
@JuanMaRuiz I consulted this case with the team. We're not sure if that workaround is needed but just in case please don't change array accesses to this new format. You can disable the rule that enforces that in the ESLint config override for `sizzle.js`.

mgol commented on src/sizzle.js:
```diff
@@ -8,1150 +8,1198 @@
-			}
+					// Choose the first element that is related to our preferred document
+					if ( a === document || a.ownerDocument === preferredDoc &&
+						contains( preferredDoc, a ) ) {
```
I'd prefer breaking the line after `||` as `&&` has precendence so it'd be simpler to understand.

JuanMaRuiz replied:
Agree

mgol commented on src/sizzle.js:
```diff
@@ -8,1150 +8,1198 @@
+						contains( preferredDoc, a ) ) {
+						return -1;
+					}
+					if ( b === document || b.ownerDocument === preferredDoc &&
```
Same remark: I'd prefer breaking the line after `||` as `&&` has precendence so it'd be simpler to understand.

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
-			}
+					// numeric x and y parameters for Expr.filter.CHILD
+					// remember that false/true cast respectively to 0/1
+					match[ 4 ] = +( match[ 4 ] ? match[ 5 ] + ( match[ 6 ] || 1 ) : 2 *
```
This is a ternary. Can you break lines after `?` & `:` instead? Currently it's hardly readable.

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
+			"PSEUDO": function( match ) {
+				var excess,
+					unquoted = !match[ 6 ] && match[ 2 ],
+					unqotedLength = unquoted.length;
```
Please don't cache this variable. This PR shouldn't change code in Sizzle source, only formatting.

JuanMaRuiz replied:
You are right

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
+					classCache( className, function( elem ) {
+						return pattern.test( typeof elem.className === "string" && elem.className ||
+							typeof elem.getAttribute !== "undefined" &&
+							elem.getAttribute( "class" ) || "" );
```
I'd prefer the following formatting if possible:
```js
return pattern.test(
	typeof elem.className === "string" && elem.className ||
	typeof elem.getAttribute !== "undefined" && elem.getAttribute( "class" ) ||
	""
);
```

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
-				forward = type.slice( -4 ) !== "last",
-				ofType = what === "of-type";
+					result += "";
+					/* eslint-disable max-len*/
```
Please add a space before the end of this comment.

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
+										operator === "~=" ? ( " " + result.replace( rwhitespace, " " ) + " " ).indexOf( check ) > -1 :
+											operator === "|=" ? result === check || result.slice( 0, check.length + 1 ) === check + "-" :
+												false;
+					/* eslint-enable max-len*/
```
Please add a space before the end of this comment.

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
+
+			"focus": function( elem ) {
+				return elem === document.activeElement && ( !document.hasFocus ||
+					document.hasFocus() ) && !!( elem.type || elem.href || ~elem.tabIndex );
```
Can we break it after each `&&` instead?

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
+					i = length;
+				} else {
+					i = argument;
+				}
```
Is this change necessary?

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
-setFilters.prototype = Expr.filters = Expr.pseudos;
-Expr.setFilters = new setFilters();
+	function setFilters() {
+	}
```
Is this change necessary?

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
-			postFilter( temp, [], context, xml );
+				// Get initial elements from seed or context
+				elems = seed || multipleContexts( selector || "*",
+					context.nodeType ? [ context ] : context, [] ),
```
Can we break it like that instead?
```js
elems = seed || multipleContexts(
	selector || "*",
	context.nodeType ? [ context ] : context,
	[]
),
```

JuanMaRuiz replied:
You are right, much better

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
+						[] :
+
+						// ...otherwise use results directly
+						results : matcherIn;
```
Can we keep the old formatting here?

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
-					j < len && matcherFromTokens( (tokens = tokens.slice( j )) ),
-					j < len && toSelector( tokens )
-				);
+						tokens.slice( 0, i - 1 ).concat( { value: tokens[ i - 2 ].type === " " ? "*" : "" } ) // eslint-disable-line max-len
```
Let's not disable the rule here, this line can be formatted better.

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
 
-	results = results || [];
+			// Cache the compiled function
+			cached = compilerCache( selector, matcherFromGroupMatchers( elementMatchers, setMatchers ) ); // eslint-disable-line max-len
```
Let's not disable the rule here, this line can be formatted better.

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
+			if ( tokens.length > 2 && ( token = tokens[ 0 ] ).type === "ID" &&
+				context.nodeType === 9 && documentIsHTML && Expr.relative[ tokens[ 1 ].type ] ) {
+
+				context = ( Expr.find.ID( token.matches[ 0 ].replace( runescape, funescape ), context ) || [] )[ 0 ]; // eslint-disable-line max-len
```
Let's not disable the rule here, this line can be formatted better.

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
-			selector = selector.slice( tokens.shift().value.length );
-		}
+			// Fetch a seed set for right-to-left matching
+			i = matchExpr.needsContext.test( selector ) ? 0 : tokens.length;
```
This may also need to be reverted if it turns out Google Closure Compiler is a concern.

JuanMaRuiz replied:
Reverted

JuanMaRuiz replied:
I have reverted the change. I also added the eslint annotation to skip this line because eslint is configured to mark as error the use of bracket notation instead of dot notation.

mgol commented on src/sizzle.js:
```diff
@@ -1161,1122 +1209,1181 @@ Expr = Sizzle.selectors = {
+					// Search, expanding context for leading sibling combinators
+					if ( ( seed = find(
+						token.matches[ 0 ].replace( runescape, funescape ),
+						rsibling.test( tokens[ 0 ].type ) && testContext( context.parentNode ) ||
```
Is this change necessary? It should fit in 100 characters.

mgol commented on Gruntfile.js:
```diff
@@ -133,43 +133,36 @@ module.exports = function( grunt ) {
-		},
-		jscs: {
+
+			// We have to explicitly declare "src" property otherwise "newer"
```
This comment doesn't apply as Sizzle is not using the `newer` task.

JuanMaRuiz commented on .eslintrc-browser.json:
```diff
@@ -0,0 +1,31 @@
+	"jQuery": true,
+	"define": true,
+	"module": true,
+	"noGlobal": true
```
I'll remove it!! thanks!

JuanMaRuiz commented on .eslintrc-browser.json:
```diff
@@ -0,0 +1,31 @@
+
+  "extends": "jquery",
+
+  // Support: IE <=9 only, Android <=4.0 only
```
I didn't notice that! Thanks

JuanMaRuiz commented:
> Thanks for the PR! I left some comments to address.

Many thanks to you for the quick review!! I have applied almost all the request. If you have any time, could to please review it again?

Best regards.

JuanMaRuiz commented:
> Thanks for the PR! I left some comments to address.

Hi @mgol I have noted that I missed the configuration of some test and tasks files. I have tried to mimic the jquery configuration. I hope the configuration is the correct ones.

Please, if you have any time let me know if everything is ok now.

Best regards.

JuanMa.

mgol commented on tasks/version.js:
```diff
@@ -1,8 +1,7 @@
 var exec = require( "child_process" ).exec;
 
 module.exports = function( grunt ) {
+	"use strict";
```
This is a Node.js file, those are better served by a global `"use strict"`; pragma. We should have a separate config for Node.js scripts & the browser files, see e.g. how https://github.com/jquery/jquery has that configured.

JuanMaRuiz replied:
You are right @mgol , my bad. I'm going to fix all task files.

Many thanks

JuanMaRuiz replied:
Removed all `use strict` from all task files. I have had to add some extra files and configuration in the `Gruntfile` to have a configuration like jQuery project. I hope now the configuration is the correct one.

mgol commented on src/sizzle.js:
```diff
@@ -8,2275 +8,2393 @@
-						context.removeAttribute( "id" );
+						// Expand context for sibling selectors
+						newContext = rsibling.test( selector ) &&
+							testContext( context.parentNode ) || context;
```
If you're putting this in a new line, I'd put the `|| context` part in a new line as well to not confuse evaluation precedence. But please re-check if it doesn't fit after de-indenting the file.

JuanMaRuiz replied:
I'll review after reindent the whole file.

JuanMaRuiz replied:
Hi @mgol finally I re-indented the `sizzle.js` file. About `|| context`, according with the `eslint` configuration the operator `||` have to be placed in the parent line. Is ok for you?

mgol replied:
Yes, that sounds OK. I added such a comment in the new review so that it doesn't get lost in outdated topics, I'm resolving this one.

mgol commented on .eslintrc-browser.json:
```diff
@@ -0,0 +1,34 @@
+{
+  "root": true,
```
Can you use tabs in this file? Only `package.json` should be indented by 2 spaces. You might want to copy jQuery's `.editorconfig` to Sizzle.

mgol replied:
This item is not resolved yet.

mgol commented on .eslintrc-browser.json:
```diff
@@ -0,0 +1,34 @@
+  // Support:
+  //  - Desktop: Chrome 16+, Edge 12+, Firefox 3.6+, Internet Explorer 7+1, Opera 11.6+, Safari 4.0+
+  //  - Mobile: Android 2.3+, iOS 5.1+
+  // More info: https://github.com/jquery/sizzle/wiki#-browser-support
```
This comment's purpose is not to list the supported browsers. This is a support comment which means it's supposed to explain why something was done in subsequent lines. Here, the goal is to explain why we're restricting the parser to `ecmaVersion: 3`. See the [jQuery comment](https://github.com/jquery/jquery/blob/3.4-stable/.eslintrc-browser.json#L6-L8) for comparison.

However, Sizzle supports a lot of ES3 browsers so I'd just remove that whole comment. In jQuery it was useful as since jQuery 2.0+ didn't support IE 8 & older one could think ES5 could be assumed which it couldn't. In Sizzle it's obvious only ES3 can be assumed.

mgol commented on .eslintrc.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "extends": ".eslintrc-node.json",
```
Please double-check all JSON files use tabs to indent.

JuanMaRuiz replied:
Done. Many thanks

mgol replied:
Some files are still not indented via tabs, e.g. `.eslintrc-browser.json`.

mgol replied:
I've added another comment about that so I'm resolving this thread.

mgol commented on Gruntfile.js:
```diff
@@ -133,42 +133,30 @@ module.exports = function( grunt ) {
+				src: files.source
+			},
+			dev: {
+				src: [ files.source, files.grunt, files.karma, files.speed, files.tests ]
```
Why is `files.source` listed twice, in `src` & `dev` sections?

JuanMaRuiz replied:
Please, check the new configuration. I have tried to do the same as jQuery

mgol commented on dist/.eslintrc.json:
```diff
@@ -0,0 +1,9 @@
+	// That is okay for the built version
+	"no-multiple-empty-lines": "off"
+  }
+}
```
I don't think this file is used right now as we don't lint `dist/sizzle.js`. Either that needs to be added to Gruntfile (as it is in jQuery) together with the modifications in `.eslintignore` that I listed above or this file needs to be removed.

JuanMaRuiz replied:
I have updated `eslint` task in the `Gruntfile` and now this file should be used

mgol commented on tasks/commit.js:
```diff
@@ -1,10 +1,9 @@
-"use strict";
```
Don't remove the global `use strict`. We shouldn't turn files that were already in strict mode into sloppy mode.

I'd also expect it to be validated by ESLint so that all the Node.js-based files do have a global `"use strict";` pragma.

JuanMaRuiz replied:
Ok, done

mgol commented on test/.eslintrc:
```diff
@@ -0,0 +1,40 @@
+		"lines-around-comment": "off",
+		"dot-notation": "off"
+	},
+}
```
Please always end files with an empty line.

mgol commented on src/sizzle.js:
```diff
@@ -8,8 +8,8 @@
-(function( window ) {
-
+( function( window ) {
+"use strict";
```
This still needs to be removed.

mgol commented on src/sizzle.js:
```diff
@@ -630,55 +654,55 @@ setDocument = Sizzle.setDocument = function( node ) {
 	// ID filter and find
 	if ( support.getById ) {
-		Expr.filter["ID"] = function( id ) {
+		Expr.filter.ID = function( id ) {
```
Let's revert that to the `[ "ID" ]` syntax. Please override the proper rule from the jquery preset in the ESLint config.

This was done for some Google Closure Compiler compatibility and while we're not verifying it in any way, we'd rather avoid the risk at this point.

This applies to here & all other changes like that done on `Expr`.

JuanMaRuiz replied:
Ok, it's done a I have added `"dot-notation": "off"` to `.eslintrc-browser.json` to override that rule.

mgol commented on src/sizzle.js:
```diff
@@ -1161,49 +1207,55 @@ Expr = Sizzle.selectors = {
 				// advance to the next closing parenthesis
-				(excess = unquoted.indexOf( ")", unquoted.length - excess ) - unquoted.length) ) {
+				( excess = unquoted.indexOf( ")",
+					unquoted.length - excess ) - unquoted.length ) ) {
```
Can you move `unquoted.length - excess ) -` to the previous line?

JuanMaRuiz replied:
Yep, done

mgol commented on src/sizzle.js:
```diff
@@ -1226,10 +1280,16 @@ Expr = Sizzle.selectors = {
+						className, function( elem ) {
+							return pattern.test(
+								typeof elem.className === "string" &&
+								elem.className || typeof elem.getAttribute !== "undefined" &&
```
Please move `elem.className ||` to the previous line so that it's clearer what is grouped with what (as `&&` takes precedence before `||`).

The same remark to the line below - move `elem.getAttribute( "class" ) ||` to this line.

JuanMaRuiz replied:
Do yo think it looks better in this way? 

![PR-improvement](https://user-images.githubusercontent.com/1837650/58383326-9a992480-7fd5-11e9-92d4-1fe052f74df4.png)

mgol replied:
@JuanMaRuiz In this way to break the second part of the expression:
```js
typeof elem.getAttribute !== undefined && elem.getAttribute( "class" )
```
is broken into two lines and the second part is put in the same line as the third condition: `""`. That's confusing to me. As I wrote above, I'd prefer each part to be kept in one line.

mgol replied:
If it's not possible to fit in the `elem.getAttribute( "class" )` part in the previous line, I'd move the last `""` to a new line to separate it from the previous condition. In that case, the line with `elem.getAttribute( "class" )` should be indented more:
```js
								typeof elem.className === "string" && elem.className ||
								typeof elem.getAttribute !== "undefined" &&
									elem.getAttribute( "class" ) ||
								""
```

JuanMaRuiz replied:
Done

mgol commented on src/sizzle.js:
```diff
@@ -1245,14 +1305,18 @@ Expr = Sizzle.selectors = {
 
 				result += "";
 
+				/* eslint-disable max-len */
```
I think the following `return` expression was clearer originally without all this extra intendation. Can you revert that to the original indentation? If ESLint complains, you can add an override like you did in this line - and maybe remove the `max-len` one.

JuanMaRuiz replied:
I need to maintain the `max-len` rule. I hope the indentation is better now.

mgol commented on src/sizzle.js:
```diff
@@ -1507,24 +1581,31 @@ Expr = Sizzle.selectors = {
 			var nodeName = elem.nodeName.toLowerCase();
-			return (nodeName === "input" && !!elem.checked) || (nodeName === "option" && !!elem.selected);
+			return ( nodeName === "input" &&
+				!!elem.checked ) || ( nodeName === "option" &&
```
Same remark as before in many places - please break in the places where it logically makes sense. Here, on `||`. There should be only two lines here, not three.

mgol commented on src/sizzle.js:
```diff
@@ -2154,20 +2264,23 @@ select = Sizzle.select = function( selector, context, results, seed ) {
+				if ( ( seed = find(
+					token.matches[ 0 ].replace( runescape, funescape ),
+					rsibling.test( tokens[ 0 ].type ) &&
+					testContext( context.parentNode ) || context
```
If this doesn't fit in the previous line, can you break after `||` as well to not suggest that `context` is a part of the same check as the `testContext` call is?

JuanMaRuiz commented:
> Thanks for the update! I added more comments.

Many thanks for your quick review @mgol ! I hope is everything is fits the requirements now.

Best regards!

mgol commented on .eslintrc-node.json:
```diff
@@ -4,7 +4,7 @@
 
 	"parserOptions": {
-		"ecmaVersion": 2017
+		  "ecmaVersion": 2017
```
Some rogue spaces got in here.

JuanMaRuiz replied:
Removed. Now there is only tabs ...

mgol commented on .eslintrc.json:
```diff
@@ -1,6 +1,6 @@
-  "rules": {
-	"no-nested-ternary": 1
-  }
+  	"extends": ".eslintrc-node.json",
```
This still doesn't use tabs consistently. It uses both spaces & tabs from what I see.

Can you configure your IDE to highlight tabs & spaces so that you can see where it went wrong?

Please re-check all the new configuration files again.

JuanMaRuiz replied:
Yep, after added `.editorconfig` file and showed indentation on my IDE I got my mistake...Sorry about that. 

Please re-check all files, they should be ok, now.

mgol commented on dist/.eslintrc.json:
```diff
@@ -1,9 +1,22 @@
+		"no-multiple-empty-lines": "off",
+
+	  // Sizzle is not compatible with jQuery code style
+	  "no-nested-ternary": "off",
```
A general remark - why do we have to disable all of these rules for `dist` when we don't do that for `src` and `sizzle.js` is just copied to `dist`? That shouldn't be necessary.

(You'll need to run `npm run build` & commit the changed files in `dist`, though)

JuanMaRuiz replied:
It was my mistake, I didn't commit my `dist` files and travis job failed. Now I have commit all the files all this configuration should be not needed.

mgol commented on dist/.eslintrc.json:
```diff
@@ -1,9 +1,22 @@
+	  "computed-property-spacing": "off",
+	  "max-len": "off",
+	  "dot-notation": "off",
+	  "semi-spacing": "off",
```
This should be left on.

JuanMaRuiz replied:
Check previous commit about this unnecessary configuration.

mgol commented on dist/.eslintrc.json:
```diff
@@ -1,9 +1,22 @@
+	  "max-len": "off",
+	  "dot-notation": "off",
+	  "semi-spacing": "off",
+	  "brace-style": "off"
```
This should be left on.

JuanMaRuiz replied:
Check previous commit about this unnecessary configuration

mgol commented on src/sizzle.js:
```diff
@@ -2278,8 +2274,8 @@ select = Sizzle.select = function( selector, context, results, seed ) {
-					rsibling.test( tokens[ 0 ].type ) &&
-					testContext( context.parentNode ) || context
+					rsibling.test( tokens[ 0 ].type ) && testContext( context.parentNode ) ||
+					context
```
Please indent that line one level more.

JuanMaRuiz replied:
Done

JuanMaRuiz commented on dist/.eslintrc.json:
```diff
@@ -1,9 +1,22 @@
+
+	  // Sizzle is not compatible with jQuery code style
+	  "no-nested-ternary": "off",
+	  "no-unused-expressions": "off",
```
I have removed this configuration because is not needed.

JuanMaRuiz commented on dist/.eslintrc.json:
```diff
@@ -1,9 +1,22 @@
+	  // Sizzle is not compatible with jQuery code style
+	  "no-nested-ternary": "off",
+	  "no-unused-expressions": "off",
+	  "no-unused-vars": "off",
```
I have removed this configuration because is not needed.

JuanMaRuiz commented on dist/.eslintrc.json:
```diff
@@ -1,9 +1,22 @@
+	  "no-nested-ternary": "off",
+	  "no-unused-expressions": "off",
+	  "no-unused-vars": "off",
+	  "lines-around-comment": "off",
```
I have removed this configuration because is not needed.

JuanMaRuiz commented:
> Thanks for the updates! See a few more comments, including in old threads that are not resolved yet (even the outdated ones).

Many thanks @mgol . I think I have done all the changes:

- Fixed all my mistakes.
- Added all minor changes you requested about indentation.

Best Regards.

mgol commented on speed/.eslintrc.json:
```diff
@@ -0,0 +1,10 @@
+		"console":  true,
+		"require": true,
+		"setTimeout": true,
+		"location": true,
```
Could you remove the `location` & `setTimeout` entries from here & change them in `speed/*` code to `window.location` & `window.setTimeout`? That'd improve compatibility with non-browser environments.

Let me know if that's too much work.

JuanMaRuiz replied:
No problem. Done

mgol commented on src/sizzle.js:
```diff
@@ -1161,49 +1206,54 @@ Expr = Sizzle.selectors = {
+				unquoted = !match[ 6 ] && match[ 2 ];
 
-			if ( matchExpr["CHILD"].test( match[0] ) ) {
+			if ( matchExpr.CHILD.test( match[ 0 ] ) ) {
```
Let's change that to `[ "CHILD" ]`.

mgol commented on src/sizzle.js:
```diff
@@ -1546,7 +1625,7 @@ Expr = Sizzle.selectors = {
 
 		"parent": function( elem ) {
-			return !Expr.pseudos["empty"]( elem );
+			return !Expr.pseudos.empty( elem );
```
Let's change that to `[ "empty" ]`.

mgol commented on src/sizzle.js:
```diff
@@ -1612,19 +1692,19 @@ Expr = Sizzle.selectors = {
 };
 
-Expr.pseudos["nth"] = Expr.pseudos["eq"];
+Expr.pseudos.nth = Expr.pseudos.eq;
```
Let's quote those properties again.

JuanMaRuiz replied:
My bad, I forgot this dot notation

mgol commented on src/sizzle.js:
```diff
@@ -188,29 +204,32 @@ var i,
-		function( target, els ) {
-			push_native.apply( target, slice.call(els) );
-		} :
+			// Leverage slice if possible
```
Can this whole block be de-indented?

JuanMaRuiz replied:
Done

mgol commented on src/sizzle.js:
```diff
@@ -262,13 +281,13 @@ function Sizzle( selector, context, results, seed ) {
 						}
 
-					// Element context
+						// Element context
```
Can you de-indent that comment?

mgol commented on src/sizzle.js:
```diff
@@ -277,13 +296,13 @@ function Sizzle( selector, context, results, seed ) {
 
-				// Type selector
-				} else if ( match[2] ) {
+					// Type selector
```
Can you de-indent that comment?

mgol commented on src/sizzle.js:
```diff
@@ -277,13 +296,13 @@ function Sizzle( selector, context, results, seed ) {
 
-				// Class selector
-				} else if ( (m = match[3]) && support.getElementsByClassName &&
+					// Class selector
```
Can you de-indent that comment?

mgol commented on src/sizzle.js:
```diff
@@ -880,100 +907,106 @@ setDocument = Sizzle.setDocument = function( node ) {
 	// Document order sorting
 	sortOrder = hasCompare ?
-	function( a, b ) {
+		function( a, b ) {
```
Can you de-indent that whole big block (both functions)? I know it looks weird but I'd just want to avoid massive indentation changes at this point so that `git blame` works as best as it could.

mgol commented on src/sizzle.js:
```diff
@@ -1161,49 +1206,54 @@ Expr = Sizzle.selectors = {
+				match[ 2 ] = match[ 4 ] || match[ 5 ] || "";
 
-			// Strip excess characters from unquoted arguments
+				// Strip excess characters from unquoted arguments
```
Can you de-indent that comment?

mgol commented on src/sizzle.js:
```diff
@@ -2137,15 +2243,16 @@ select = Sizzle.select = function( selector, context, results, seed ) {
 				return results;
 
-			// Precompiled matchers will still verify ancestry, so step up a level
+				// Precompiled matchers will still verify ancestry, so step up a level
```
Can you de-indent that comment?

mgol commented on src/sizzle.js:
```diff
@@ -2154,20 +2261,23 @@ select = Sizzle.select = function( selector, context, results, seed ) {
 
 		// Fetch a seed set for right-to-left matching
-		i = matchExpr["needsContext"].test( selector ) ? 0 : tokens.length;
+		// eslint-disable-next-line dot-notation
```
This comment is no longer needed.

mgol commented on test/.eslintrc:
```diff
@@ -0,0 +1,40 @@
+
+		// Not really too many - waiting for autofix features for these rules
+		"lines-around-comment": "off",
+		"dot-notation": "off"
```
The above comment doesn't apply to this rule as we're overriding it on purpose now. The rule should be in its own separate section, maybe with a short explanation that it's turned off for compatibility with Google Closure Compiler's advanced mode.

JuanMaRuiz replied:
You are right. Added comment

mgol commented on src/sizzle.js:
```diff
@@ -2195,88 +2305,93 @@ select = Sizzle.select = function( selector, context, results, seed ) {
-		}
-	});
-}
+	// One-time assignments
```
Please de-indent that whole block.

JuanMaRuiz commented on dist/.eslintrc.json:
```diff
@@ -6,17 +6,7 @@
-	  "dot-notation": "off",
-	  "semi-spacing": "off",
-	  "brace-style": "off"
+		// Sizzle is not compatible with jQuery code style
```
Good idea. Added!!

JuanMaRuiz commented on package.json:
```diff
@@ -64,6 +64,7 @@
     "requirejs-text": "2.0.15"
   },
   "scripts": {
+    "lint": "grunt eslint",
```
Done

JuanMaRuiz commented:
> Thanks for all the work! It's getting close. See my newest comments (all the old threads are resolved now).

Great!! I have added all the changes you requested. 

Best regards.

mgol commented:
@timmywil @gibson042 Do you want to review this as well? The PR LGTM so - if not - I'd land it soon.

mgol commented:
I compared the minified file on `master` & on this branch & I found two bugs so it's good that I checked! I fixed them, see the screenshot for changes:
![Screen Shot 2019-05-29 at 17 00 13](https://user-images.githubusercontent.com/1758366/58567961-a71ab880-8233-11e9-8ca2-0609326f6124.png)
The first issue was forgetting a `|` when breaking a string into two lines. The second one was that the commit 7d92424c1a65a367a45034803e9de97462ea42e5 got reverted. After my fixes the only difference in the minified file is the version in the header that wasn't updated before so we should have no more issues. Phew!

mgol commented:
BTW, especially the first issue means we're lacking test coverage in that area.

mgol commented:
Landed, thanks @JuanMaRuiz for the whole effort!

JuanMaRuiz commented:
> Landed, thanks @JuanMaRuiz for the whole effort!

Thanks to you @mgol for your comments and review.
